xapian-applications/omega/handler_libebook.cc

```diff
@@ -0,0 +1,133 @@
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+#include <config.h>
+#include "handler.h"
+
+#include <xapian.h>
+#include <string.h>
+#include <unordered_map>
+#include <memory>
+#include <iostream>
+
+#include <librevenge-0.0/librevenge-generators/librevenge-generators.h>
+#include <librevenge-0.0/librevenge-stream/librevenge-stream.h>
+#include <libe-book/libe-book.h>
+
+using libebook::EBOOKDocument;
+using namespace std;
+using namespace librevenge;
+
+static
+char* str_copy(const char* text) {
+    int len = strlen(text);
+    char* line = new char[len];
+    strcpy(line, text);
+    return line;
+}
+
+static
+void clear_text(string& str,const char* text) {
+    int len = strlen(text);
+    for (int i = 0; i < len; ++i)
+	if (!isspace(text[i]) || (i && !isspace(text[i - 1])))
+	    str.push_back(text[i]);
+}
+
+bool
+extract(const string& filename,
+	string& dump,
+	string& title,
+	string& keywords,
+	string& author,
+	string& pages)
+{
+    try {
+	shared_ptr<RVNGInputStream> input;
+	RVNGString content_dump, metadata_dump;
+	const char* file = filename.c_str();
+
+	if (RVNGDirectoryStream::isDirectory(file))
+	    input.reset(new RVNGDirectoryStream(file));
+	else
+	    input.reset(new RVNGFileStream(file));
+
+	EBOOKDocument::Type type = EBOOKDocument::TYPE_UNKNOWN;
+	EBOOKDocument::Confidence confidence =
+				EBOOKDocument::isSupported(input.get(), &type);
+
+	if (EBOOKDocument::CONFIDENCE_SUPPORTED_PART == confidence) {
+	    input.reset(RVNGDirectoryStream::createForParent(file));
+	    confidence = EBOOKDocument::isSupported(input.get(), &type);
+	}
+
+	if ((EBOOKDocument::CONFIDENCE_EXCELLENT != confidence) &&
+	    (EBOOKDocument::CONFIDENCE_WEAK != confidence))
+	    return false;
+
+	RVNGTextTextGenerator metadata(metadata_dump, true);
+
+	if (EBOOKDocument::RESULT_OK !=
+	    EBOOKDocument::parse(input.get(), &metadata, type))
+	    return false;
+
+	unordered_map<string,char*> hash;
+	char* str = str_copy(metadata_dump.cstr());
+	char* pch = strtok(str, "\n");
+	while (pch != NULL) {
+	    char* sep = strpbrk(pch, " ");
+	    *sep = '\0';
+	    hash[pch] = sep + 1;
+	    pch = strtok(NULL, "\n");
+	}
+	delete[] str;
+	unordered_map<string,char*>::const_iterator it;
+
+	// Extract Author if it possible
+	if ((it = hash.find("dc:creator")) != hash.end())
+	    clear_text(author, it->second);
+	else if ((it = hash.find("meta:initial-creator")) != hash.end())
+	    clear_text(author, it->second);
+
+	// Extract Title if it possible
+	if ((it = hash.find("dc:title")) != hash.end())
+	    clear_text(title, it->second);
+
+	// Extract Keywords if it possible
+	if ((it = hash.find("dc:subject")) != hash.end())
+	    clear_text(keywords, it->second);
+	if ((it = hash.find("dcterms:available")) != hash.end())
+	    clear_text(keywords, it->second);
+	pages = "";
```

Comment:
I think we should just rely on the caller arranging for these variables to be empty before calling us (as the current implementation does).  That way we avoid having to have code in each module to clear any fields which aren't handled - instead that's all dealt with once in one place in the calling code in `assistant.cc`.

Reply:
Actually, if I don't use an argument I get a compiler error. That is why I assign an empty string to pages.

Reply:
What was the compiler error?

If it's "unused parameter" the answer is to tell the compiler that's OK.  There are various ways to do this:

* remove the parameter name from the method definition (so it says `..., string& author, string&)`) - this is portable, but only works for C++ (C doesn't allow omitting parameter names in function definitions).  This has the downside that it's not clear what the parameter is without looking at the header, though you can comment it out instead of removing it like so: `string& /*pages*/`

* add a dummy use via `(void)pages;` (conventionally at the start of the function) which is a way to tell the compiler that it's OK that the parameter is otherwise unused.  This is portable and most compilers understand it as meaning "pretend `pages` is used".  This is the probably the best option in portable C code.

* C++17 added a standard attribute (https://en.cppreference.com/w/cpp/language/attributes/maybe_unused) but currently Xapian only requires a C++11 compiler.

* There are also compiler-specific ways to specify this (e.g. `__attribute__((unused))` for GCC and clang), but those are inherently less portable.

So here I'd go for the `(void)` cast or commenting out the parameter name.

Reply:
Sorry, I didn't know that. I am going to fix it.